R/create_individual_file.R

```diff
@@ -0,0 +1,924 @@
+        "{prefix}_daycase_episodes" := dplyr::if_else(eval(condition_d), 1, NA_real_),
+        "{prefix}_daycase_cost" := dplyr::if_else(eval(condition_d), .data$cost_total_net, NA_real_)
+      )
+  }
+  return(episode_file)
+}
+
+#' Add standard columns
+#'
+#' @description Add standard columns (DoB, postcode, gpprac, episodes, cost) to episode file.
+#'
+#' @inheritParams add_acute_columns
+#' @param drop Any columns out of "DoB", "postcode", and "gpprac" that should be dropped
+#' @param episode Whether to create prefix_episodes col, e.g. "Acute_episodes"
+#' @param cost Whether to create prefix_cost col, e.g. "Acute_cost"
+add_standard_cols <- function(episode_file, prefix, condition, drop = NULL, episode = FALSE, cost = FALSE) {
+  episode_file <- dplyr::bind_cols(episode_file, create_cols(episode_file, prefix, condition, drop))
+  if (episode) {
+    episode_file <- dplyr::mutate(episode_file, "{prefix}_episodes" := dplyr::if_else(eval(condition), 1, NA_real_))
+  }
+  if (cost) {
+    episode_file <- dplyr::mutate(episode_file, "{prefix}_cost" := dplyr::if_else(eval(condition), .data$cost_total_net, NA_real_))
+  }
+  return(episode_file)
+}
+
+#' Create standard cols
+#'
+#' @description Create standard cols (DoB, postcode, gpprac).
+#'
+#' @inheritParams add_acute_columns
+#' @param drop Any columns out of "DoB", "postcode", and "gpprac" that should be dropped
+create_cols <- function(episode_file, prefix, condition, drop) {
+  cols <- c("DoB", "postcode", "gpprac")
+  if (!is.null(drop)) {
+    cols <- cols[cols != drop]
+  }
+  episode_file <- purrr::map_dfc(cols, ~ create_col(episode_file, .x, prefix, condition))
+  return(episode_file)
+}
+
+#' Create standard col
+#'
+#' @description Create single standard column.
+#'
+#' @inheritParams add_acute_columns
+#' @inheritParams na_type
+create_col <- function(episode_file, col, prefix, condition) {
+  episode_file %>%
+    dplyr::mutate("{prefix}_{col}" := dplyr::if_else(eval(condition), .data[[tolower(col)]], na_type(col))) %>%
+    dplyr::select(dplyr::last_col())
+}
+
+#' NA type
+#'
+#' @description Helper function to use correct NA type depending on
+#' which type of column is created.
+#'
+#' @param col Which column to create ("DoB", "postcode", or "gpprac")
+na_type <- function(col = c("DoB", "postcode", "gpprac")) {
+  match.arg(col)
+  na_type <- switch(col,
+    "DoB" = lubridate::NA_Date_,
+    "postcode" = NA_character_,
+    "gpprac" = NA_real_
+  )
+  return(na_type)
+}
+
+#' Aggregate CIS episodes
+#'
+#' @description Aggregate CH variables by CHI and CIS.
+#'
+#' @inheritParams create_individual_file
+aggregate_ch_episodes <- function(episode_file) {
+  episode_file %>%
+    dplyr::group_by(.data$chi, .data$ch_chi_cis) %>%
+    dplyr::mutate(
+      ch_no_cost = max(.data$ch_no_cost),
+      ch_ep_start = min(.data$record_keydate1),
+      ch_ep_end = max(.data$ch_ep_end),
+      ch_cost_per_day = mean(.data$ch_cost_per_day)
+    ) %>%
+    dplyr::ungroup()
+}
```

Comment:
Split out and do filter on records with a chi and ch_chi_cis

Something like this might be faster but maybe not... would need testing! There's possibly a way to do this without any grouping?
```suggestion
aggregate_ch_episodes <- function(episode_file) {
  episode_file %>%
  dplyr::filter(!is.na(ch_chi_si) %>%
    dplyr::group_by(.data$chi, .data$ch_chi_cis) %>%
    dplyr::mutate(
      ch_no_cost = max(.data$ch_no_cost),
      ch_ep_start = min(.data$record_keydate1),
      ch_ep_end = max(.data$ch_ep_end),
      ch_cost_per_day = mean(.data$ch_cost_per_day)
    ) %>%
    dplyr::ungroup() %>% 
    dplyr::distinct(.data$chi, .data$ch_chi_cis) %>%
    dplyr::select(.data$chi, .data$ch_chi_cis, ch_no_cost, ...) %>%
    dplyr::right_join(episode_file, by = c(.data$chi, .data$ch_chi_cis))
}
```
Other suggestions use the following for speeding up code: 
as.datatable()
multidplyr (run commands in parallel)